### DISCUSSION TITLE
Email form using Sendgrid nextjs13

### DISCUSSION THREAD

**Discussion by Vukichinuky** [vercel/next.js#47969] [Help]
### Summary

 POST http://localhost:3001/api/sendEmail 405 (Method Not Allowed) this is error i get..
 this code is ContactForm.tsx:
`import { useState } from "react";

const ContactForm = () => {
    const [formData, setFormData] = useState({ email: "", message: "" });
    const [formMessage, setFormMessage] = useState("");

    const handleSubmit = async (e: { preventDefault: () => void; }) => {
        e.preventDefault();
        try {
            const res = await fetch("/api/sendEmail", {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify(formData),
            });
            const data = await res.json();
            setFormMessage(data.message);
            setFormData({ email: "", message: "" });
        } catch (error) {
            console.error(error);
            setFormMessage("Something went wrong.");
        }
    };

    const handleChange = (e: { target: { name: any; value: any; }; }) => {
        setFormData({ ...formData, [e.target.name]: e.target.value });
    };

    return (
        <form onSubmit={handleSubmit}>
            <div>
                <label htmlFor="email">Email:</label>
                <input
                    type="email"
                    id="email"
                    name="email"
                    value={formData.email}
                    onChange={handleChange}
                    required
                />
            </div>
            <div>
                <label htmlFor="message">Message:</label>
                <textarea
                    id="message"
                    name="message"
                    value={formData.message}
                    onChange={handleChange}
                    required
                ></textarea>
            </div>
            <button type="submit">Send</button>
            {formMessage && <p>{formMessage}</p>}
        </form>
    );
};

export default ContactForm;
`

this is app/api/sendEmail/route.ts :`import { NextApiRequest, NextApiResponse } from "next";
import sgMail from "@sendgrid/mail";
require("dotenv").config();
const { SENDGRID_API_KEY } = process.env;
sgMail.setApiKey(SENDGRID_API_KEY as string);
console.log(process.env.SENDGRID_API_KEY);

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const { email, message } = req.body;

  const msg = {
    to: "drvuk23@gmail.com", // Replace with your own email address
    from: email,
    subject: "New message from your website",
    text: message,
  };

  try {
    await sgMail.send(msg);
    res.status(200).json({ message: "Email sent successfully!" });
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: "Something went wrong." });
  }
}
`

and this is .env.local SENDGRID_API_KEY=someapi

### Additional information

```js
POST http://localhost:3001/api/sendEmail 405 (Method Not Allowed) this is error i get..
```

### Example

405 (Method Not Allowed) this is error i get..

**Comment by haydenbleasel** [comment]
@Vukichinuky Haven't run this locally but it may be because you've written this as a Next API Handler, however it's in a `route.ts` file which signifies it's meant to be written as a [Route Handler](https://beta.nextjs.org/docs/routing/route-handlers). Route Handlers have their accepted HTTP methods defined as different exports e.g. `export async function GET(request: Request) {}` rather than a single default exported `handler` function.

**Reply by icyJoseph** [reply]
Ah yeah that must be the problem. I had missed that...
